`sticky` LoadBalancer implementation (#274)
smallrye/smallrye-stork: docs/load-balancer/sticky.md

# Sticky Load Balancing

The sticky load balancer selects a single service instance and keeps using it until it fails.

> Then, it selects the next one.

Reviewer: instead of `next`, I would say `another` one.

> It is possible to configure a back off time to specify for how long a service instance that failed should not be retried.

Reviewer: ```suggestion
It is possible to configure a backoff time to specify for how long a failing service instance should not be retried.
```

Precisely, the load balancer works as follows:

> * if the previously selected service instance has not failed, and is still available, return it;

Reviewer: Maybe add the initial state: if no instance has been selected, select the first instance from the list.
(actually, random may be good too).

> * if the previously selected service instance has not failed, and is still available, return it

Reviewer: ```suggestion
* if the previously selected service instance has not failed, and is still available, return it;
```

> * else if exists, return the first available service instance that has no recorded failure;

Reviewer: I would reword the if exists: if the instance is still available, but has failed a previous invocation...

> * else if exists, return the first available service instance that has no recorded failure

Reviewer: ```suggestion
* else if the service instance is still available, but has failed during a previous invocation, return the first available service instance that has no recorded failure;
```

Author: the corrected sentences is not true. If no service was selected previously, we should fall under this branch too. Also it may happen that the previously selected instance did not fail but is just not availalbe anymore (service discovery-wise)

> * else, find the available instance that has failed the longest time ago, and

Reviewer: ```suggestion
* otherwise, find the available instance that was the first to record a failure, and
```

Author: that's not technically true. 
Let's assume we have services A and B.
A failed first, then B failed while backoff time had passed for A. So we used A. Then A failed.
At this point we don't select A.

Author: I reworded it a bit

> * else, find the available instance for which the time since the last failure is the greatest, and

Reviewer: ```suggestion
* else, find an available instance for which the time since the last failure is the longest, and
```

Author: Why `an` and not `the`? We're talking about a specific instance here.

>     * if the back-off time since failure passed, return it

Reviewer: ```suggestion
    * if the backoff time since the failure passed, return it;
```

>     * else, report lack of applicable services (throw `NoAcceptableServiceInstanceFoundException`)

Reviewer: ```suggestion
    * or, throw an `NoAcceptableServiceInstanceFoundException` as no acceptable instances are available.
```

## Configuration

To use the `sticky` load service selection strategy, set the load balancer type to `sticky`:

```properties
stork.my-service.service-discovery=...
stork.my-service.service-discovery...=...
stork.my-service.load-balancer.type=sticky
```

The following attributes are supported:

> --8<-- "load-balancer/least-response-time/target/classes/META-INF/stork-docs/least-response-time-lb-attributes.txt"

Reviewer: That looks wrong...
